### DISCUSSION TITLE
Proper way to install language server for Jupyterlab

### DISCUSSION THREAD

**Discussion by Pilgrim1379** [jupyter/notebook#7213] [Q&A]
I'm new to using Jupyterlab/Jupyter notebook. What's baffling me is, I have installed all the language servers that I need however Jupyterlab doesn't find them. I'm therefore wondering if there's a proper way to install language servers for Jupyterlab to find and use them.

My setup is as follows:
I'm on a MacOS (latest version) and I installed Jupyterlab via pipx with the following command:
```zsh
            pipx install ruff
            pipx install ruff-lsp
            pipx install jupyterlab python-lsp-server && \
                pipx inject jupyterlab python-lsp-ruff 
            pipx install jupyterlab --include-deps && \
                pipx inject jupyterlab jupyterlab-lsp && \
                pipx inject jupyterlab jupyter-lsp && \
                pipx inject jupyterlab jupyterlab-git && \
                pipx inject jupyterlab jupyterlab_code_formatter && \
                pipx inject jupyterlab jupyterlab_templates && \
                pipx inject jupyterlab ipywidgets && \
                pipx inject jupyterlab catppuccin-jupyterlab && \
                pipx inject jupyterlab catppuccin-matplotlib
```

I have other language servers installed via npm:
```zsh
npm install -g typescript-language-server
npm install -g bash-language-server
npm install -g vscode-json-languageservice
npm install -g yaml-language-server
npm install -g diagnostic-languageserver
npm install -g pyright
npm install -g emmet-ls
npm install -g vscode-markdown-languageserver
```
In a virtual env, I install `ipykernel` and then `python -m ipykernel install --user --display-name ${PWD} --name ${PWD##*/}` to create a kernel for the virutal environment. With this I'm able to start and use Jupyterlab in the virtual environment.

When I start Jupyterlab, it doesn't find python-language-server or vscode-markdown-languageserver.

Will appreaciate any advice or pointers.

**Comment by jtpio** [comment]
Thanks @Pilgrim1379.

Would you mind opening the discussion in the JupyterLab repo, as it will likely have more changes to be seen by JupyterLab developers: https://github.com/jupyterlab/jupyterlab/discussions

Thanks!

**Reply by Pilgrim1379** [reply]
Ok will do thanks.

**Comment by Eyobs-droid** [comment]
Okay so once the jupyter lab is installed enable the jupyter lab language server manager extension in jupyter lab. To do this first launch the jupyter lab by running the following command "jupyter lab" then go to settings menu and select advanced settings editor, then select the jupyter lab configuration on the left-hand side. Then in the right-hand pane, scroll down until you find the "language_servers" section. Then add the configuration for each language server you want to use. 

 Eg:  {
        "language_servers": {
        "python": {
          "language": "python",
          "command": "python", 
          "args": ["-m", "jedi_language_server"]
       }
     }
   }


After adding the configuration tap the save button in the top menu and then close and restart the jupyter lab
